tests/framework/prototype.py

```diff
@@ -0,0 +1,279 @@
+    os.system("kubectl create -k \"github.com/ray-project/kuberay/manifests/cluster-scope-resources?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def download_images(images):
+    """Download Docker images from DockerHub"""
+    docker_client = docker.from_env()
+    for image in images:
+        docker_client.images.pull(image)
+    docker_client.close()
+
+def kind_load_images(images):
+    """Load downloaded images into KinD cluster"""
+    for image in images:
+        os.system(f'kind load docker-image {image}')
+
+def install_kuberay_operator():
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl apply -k \"github.com/ray-project/kuberay/manifests/base?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def check_cluster_exist():
+    return os.system("kubectl cluster-info --context kind-kind") == 0 
+
+'''
+Configuration Test Framework Abstractions: (1) Mutator (2) Rule (3) RuleSet (4) CREvent
+'''
+
+# Mutator: Mutator will start to mutate from `baseCR`. `patch_list` is a list of JsonPatch, and you can
+#          specify multiple fields that want to mutate in a single JsonPatch.
+class Mutator:
+    def __init__(self, baseCR, patch_list: list[jsonpatch.JsonPatch]):
+        self.baseCR = baseCR
+        self.patch_list = patch_list
+    # Generate a new cr by applying the json patch to `cr`. 
+    def mutate(self):
+        for patch in self.patch_list:
+            yield patch.apply(self.baseCR)
+
+# Rule: Rule is used to check whether the actual cluster state is the same as our expectation after a CREvent.
+#       We can infer the expected state by CR YAML file, and get the actual cluster state by Kubernetes API.
+# Example: "HeadPodNameRule"
+class Rule:
+    def __init__(self):
+        pass
+    # The rule will only be checked when `trigger_condition` is true. For example, we will only check
+    # "HeadPodNameRule" when "spec.headGroupSpec" is defined in CR YAML file.
+    def trigger_condition(self, cr=None) -> bool:
+        return True
+    def assertRule(self, cr=None, cr_namespace='default'):
+        raise NotImplementedError
+
+# RuleSet: A set of Rule
+class RuleSet:
+    def __init__(self, rules: list[Rule]):
+        self.rules = rules
+    def checkRuleSet(self, cr, namespace):
+        for rule in self.rules:
+            if rule.trigger_condition(cr):
+                rule.assertRule(cr, namespace)
+
+# CREvent: Custom Resource Event can be mainly divided into 3 categories.
+#   (1) Add (create) CR (2) Update CR (3) Delete CR
+#
+# The member functions integrate together in `trigger()`.
+#   [Step1] exec(): Execute a command to trigger the CREvent. For example, create a CR by a
+#                  `kubectl apply` command.
+#   [Step2] wait(): Wait for the system to converge.
+#   [Step3] checkRuleSets(): When the system converges, check all registered RuleSets.
+class CREvent:
+    def __init__(self, cr, ruleSets: list[RuleSet], timeout, namespace):
```

Comment:
A list of sets is a lot to think about.

Could we just have a list of Rules instead? What is the benefit of having multiple Rule sets?
If we just had list[Rule] instead, we would also be able to simplify things by removing the notion of a RuleSet.

Reply:
We will have many rules in the future and different sets of YAML files need to respect different sets of rules. The abstraction RuleSet can simplify the process. For example, we have 3 RuleSets A, B, and C, and each has 10 Rules.

YAML set 1 => RuleSet A, B, C
YAML set 2 => RuleSet B
YAML set 3 => RuleSet A, B

In addition, we can form some related rules into a RuleSet, and the RuleSet can help users understand the rules easily. For example, we can gather the rules for head pods into a RuleSet, and rules for ingress for the other RuleSet.

Reply:
Added a new section "What’s the difference between a RuleSet and a list of Rules?" to explain the understandability improvement from RuleSet.